### ISSUE TITLE
False negative: no-member in conditional branch

### ISSUE THREAD

**Issue by hemberger** [pylint-dev/pylint#9405]
### Bug description

```python
There is a difference in behavior inside conditional branches depending on if the branch is reachable or not. If it is unreachable, then `undefined-variable` warnings are emitted, but `no-member` warnings are not (whereas if it is reachable, both `undefined-variable` and `no-member` warnings _are_ emitted).

For example:

#!/usr/bin/env python3
"""False negative no-member"""
import datetime

SOME_BOOL = False
if SOME_BOOL:
    print(bar)
    print(datetime.does_not_exist)
```
This does not report `no-member` for `datetime.does_not_exist`. However, if we change `SOME_BOOL = False` to `SOME_BOOL = True`, then the `no-member` warning is emitted.

This can also happen for more complex conditionals, where it's not clear whether or not the branch is reachable. If pylint is performing fewer checks in conditional branches that it determines are unreachable as a performance optimization, then it probably should at least report that the branch as unreachable.

Thank you!
```


### Configuration

_No response_

### Command used

```shell
pylint test.py
```

### Pylint output

```python
************* Module test
test.py:7:10: E0602: Undefined variable 'bar' (undefined-variable)
```

### Expected behavior

************* Module test
test.py:7:10: E0602: Undefined variable 'bar' (undefined-variable)
test.py:8:10: E1101: Module 'datetime' has no 'does_not_exist' member (no-member)

### Pylint version

```shell
pylint 3.0.3
astroid 3.0.1
Python 3.10.13 (main, Sep 11 2023, 13:44:35) [GCC 11.2.0]
```

### OS / Environment

RHEL8

**Comment by jacobtylerwalls** [comment]
This change was made in #4764 to address false positives where the `if` guards are performing a type-narrowing function.

There are two follow-ups I see immediately from your report, though.

1. there's a better way of doing this now that we have `constraints` in `astroid`. See https://github.com/pylint-dev/astroid/pull/1189. I think your report should stand as the feature request for that.

2. Regarding:

> This can also happen for more complex conditionals, where it's not clear whether or not the branch is reachable.

We might be able to solve that by leveraging a recent feature:

```diff
diff --git a/pylint/checkers/typecheck.py b/pylint/checkers/typecheck.py
index 9c73fee82..e559a8bdd 100644
--- a/pylint/checkers/typecheck.py
+++ b/pylint/checkers/typecheck.py
@@ -513,7 +513,7 @@ def _emit_no_member(
     parent: nodes.NodeNG = node.parent
     while parent != scope:
         if isinstance(parent, (nodes.If, nodes.IfExp)):
-            inferred = safe_infer(parent.test)
+            inferred = safe_infer(parent.test, compare_constants=True)
             if (  # pylint: disable=too-many-boolean-expressions
                 isinstance(inferred, nodes.Const)
                 and inferred.bool_value() is False
```

Would you be willing to see if that solves your use case? And if so, open a modest PR attaching a functional test? Many thanks!

**Comment by hemberger** [comment]
Thanks for the response! Your suggested patch does indeed fix the more complex conditional case, though the simple case still produces a false negative. I'll see if I can create a test from the more complex case, and if so then I'll submit a PR.
